src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -946,7 +946,7 @@ public void onRefreshSettings(Settings settings) {
         @Override
         public void run() {
             // we check before if a refresh is needed, if not, we reschedule, otherwise, we fork, refresh, and then reschedule
-            if (!engine().refreshNeeded()) {
+            if (!engineSafe().refreshNeeded()) {
```

Comment:
lemme look into this but it's unrelated to this issue so I will go and handle this seperatly if you are ok with this.